File: backend/app/schemas/dev_tools.py
from typing import Literal

from pydantic import BaseModel

PokemonNumericField = Literal[
    "hp", "attack", "defense", "special_attack", "special_defense", "speed"
]
MoveNumericField = Literal["power", "accuracy", "pp", "priority", "effect_chance"]


class PokemonStatUpdate(BaseModel):
    """Dev-tool request body: directly overwrite one cached numeric stat,
    bypassing the ingestion/validation pipeline. Deliberately simulates
    "PokeAPI drifted from what's cached" so the scan's detection behavior
    can be demoed on demand rather than waiting for a real upstream change."""

    field: PokemonNumericField
    value: int


class MoveStatUpdate(BaseModel):
    field: MoveNumericField
    value: int


class MovepoolCorruption(BaseModel):
    """Dev-tool request body: add a move to a Pokemon's cached movepool that
    PokeAPI doesn't actually list for it. The next Pokemon scan re-fetches
    the real movepool, sees this move is missing from it, and treats it as
    "no longer learnable" — unassigning it from any team that has it
    equipped (if you've assigned it via the Team Builder in the meantime)
    and alerting the owner. This is how a real removed-move scenario is
    simulated, since there's no way to make PokeAPI itself drop a move on
    demand."""

    move_id: int

File: frontend/src/api/devTools.ts
import { apiFetch } from "./client";
import type { ScanResult } from "../types/scan";

export type PokemonNumericField =
  | "hp"
  | "attack"
  | "defense"
  | "special_attack"
  | "special_defense"
  | "speed";

export type MoveNumericField = "power" | "accuracy" | "pp" | "priority" | "effect_chance";

export const POKEMON_NUMERIC_FIELDS: PokemonNumericField[] = [
  "hp",
  "attack",
  "defense",
  "special_attack",
  "special_defense",
  "speed",
];

export const MOVE_NUMERIC_FIELDS: MoveNumericField[] = [
  "power",
  "accuracy",
  "pp",
  "priority",
  "effect_chance",
];

export function scanPokemon(limit?: number): Promise<ScanResult> {
  const query = limit != null ? `?limit=${limit}` : "";
  return apiFetch<ScanResult>(`/api/internal/scan-pokemon${query}`, {
    method: "POST",
    withClientId: false,
    withInternalSecret: true,
  });
}

export function scanMoves(limit?: number): Promise<ScanResult> {
  const query = limit != null ? `?limit=${limit}` : "";
  return apiFetch<ScanResult>(`/api/internal/scan-moves${query}`, {
    method: "POST",
    withClientId: false,
    withInternalSecret: true,
  });
}

export function corruptPokemonStat(
  pokemonId: number,
  field: PokemonNumericField,
  value: number,
): Promise<void> {
  return apiFetch<void>(`/api/internal/pokemon/${pokemonId}`, {
    method: "PATCH",
    body: { field, value },
    withClientId: false,
    withInternalSecret: true,
  });
}

export function corruptMoveStat(
  moveId: number,
  field: MoveNumericField,
  value: number,
): Promise<void> {
  return apiFetch<void>(`/api/internal/moves/${moveId}`, {
    method: "PATCH",
    body: { field, value },
    withClientId: false,
    withInternalSecret: true,
  });
}

/** Adds a move to a Pokemon's cached movepool that PokeAPI doesn't actually
 * list for it — the next Pokemon scan sees it's missing from the real
 * movepool and treats it as "no longer learnable" (unassigning it from any
 * team that has it equipped, and alerting the owner). This is how a
 * removed-move scenario gets simulated on demand. */
export function corruptPokemonMovepool(pokemonId: number, moveId: number): Promise<void> {
  return apiFetch<void>(`/api/internal/pokemon/${pokemonId}/movepool`, {
    method: "POST",
    body: { move_id: moveId },
    withClientId: false,
    withInternalSecret: true,
  });
}
